File: backend/pkg/api/source/handlers.go
package source

import (
	"sync"

	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	coreSource "backend/pkg/core/source"
	"backend/pkg/database"
	"backend/pkg/database/repository"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

var (
	serviceInstance *coreSource.Service
	serviceOnce     sync.Once
)

// getService returns the source service instance (lazily initialized)
func getService() *coreSource.Service {
	serviceOnce.Do(func() {
		db := database.DB()
		sourceRepo := repository.NewSourceRepository(db)
		articleRepo := repository.NewArticleRepository(db)
		serviceInstance = coreSource.NewService(sourceRepo, articleRepo)
	})
	return serviceInstance
}

// ListAllSources handles GET /dashboard/sources
// @Summary List all sources
// @Description Get a paginated list of all sources
// @Tags sources
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param limit query int false "Items per page" default(20)
// @Success 200 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /dashboard/sources [get]
func ListAllSources(c *fiber.Ctx) error {
	svc := getService()
	page := c.QueryInt("page", 1)
	limit := c.QueryInt("limit", 20)

	result, err := svc.List(c.Context(), page, limit)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"sources":     result.Sources,
		"total_pages": result.TotalPages,
		"page":        result.Page,
	})
}

// CreateSource handles POST /sources
// @Summary Create source
// @Description Create a new article source
// @Tags sources
// @Accept json
// @Produce json
// @Param request body dto.CreateSourceRequest true "Source details"
// @Success 201 {object} response.SuccessResponse{data=dto.SourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources [post]
func CreateSource(c *fiber.Ctx) error {
	svc := getService()
	var req coreSource.CreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	source, err := svc.Create(c.Context(), req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, source)
}

// ScrapeAndCreateSource handles POST /sources/scrape
// @Summary Scrape and create source
// @Description Scrape a URL and create a source from it
// @Tags sources
// @Accept json
// @Produce json
// @Param request body object{article_id=string,url=string} true "Scrape request"
// @Success 201 {object} response.SuccessResponse{data=dto.SourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources/scrape [post]
func ScrapeAndCreateSource(c *fiber.Ctx) error {
	svc := getService()
	var req struct {
		ArticleID uuid.UUID `json:"article_id" validate:"required"`
		URL       string    `json:"url" validate:"required,url"`
	}

	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	source, err := svc.ScrapeAndCreate(c.Context(), req.ArticleID, req.URL)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, source)
}

// GetArticleSources handles GET /sources/article/:articleId
// @Summary Get article sources
// @Description Get all sources for an article
// @Tags sources
// @Accept json
// @Produce json
// @Param articleId path string true "Article ID"
// @Success 200 {object} response.SuccessResponse{data=object{sources=[]dto.SourceResponse}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources/article/{articleId} [get]
func GetArticleSources(c *fiber.Ctx) error {
	svc := getService()
	articleIDStr := c.Params("articleId")
	articleID, err := uuid.Parse(articleIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid article ID"))
	}

	sources, err := svc.GetByArticleID(c.Context(), articleID)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{"sources": sources})
}

// GetSource handles GET /sources/:sourceId
// @Summary Get source
// @Description Get a source by ID
// @Tags sources
// @Accept json
// @Produce json
// @Param sourceId path string true "Source ID"
// @Success 200 {object} response.SuccessResponse{data=dto.SourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources/{sourceId} [get]
func GetSource(c *fiber.Ctx) error {
	svc := getService()
	sourceIDStr := c.Params("sourceId")
	sourceID, err := uuid.Parse(sourceIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid source ID"))
	}

	source, err := svc.GetByID(c.Context(), sourceID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, source)
}

// UpdateSource handles PUT /sources/:sourceId
// @Summary Update source
// @Description Update an existing source
// @Tags sources
// @Accept json
// @Produce json
// @Param sourceId path string true "Source ID"
// @Param request body dto.UpdateSourceRequest true "Source update details"
// @Success 200 {object} response.SuccessResponse{data=dto.SourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources/{sourceId} [put]
func UpdateSource(c *fiber.Ctx) error {
	svc := getService()
	sourceIDStr := c.Params("sourceId")
	sourceID, err := uuid.Parse(sourceIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid source ID"))
	}

	var req coreSource.UpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	source, err := svc.Update(c.Context(), sourceID, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, source)
}

// DeleteSource handles DELETE /sources/:sourceId
// @Summary Delete source
// @Description Delete a source by ID
// @Tags sources
// @Accept json
// @Produce json
// @Param sourceId path string true "Source ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources/{sourceId} [delete]
func DeleteSource(c *fiber.Ctx) error {
	svc := getService()
	sourceIDStr := c.Params("sourceId")
	sourceID, err := uuid.Parse(sourceIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid source ID"))
	}

	if err := svc.Delete(c.Context(), sourceID); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// SearchSimilarSources handles GET /sources/article/:articleId/search
// @Summary Search similar sources
// @Description Search for similar sources within an article using semantic search
// @Tags sources
// @Accept json
// @Produce json
// @Param articleId path string true "Article ID"
// @Param q query string true "Search query"
// @Param limit query int false "Max results" default(5)
// @Success 200 {object} response.SuccessResponse{data=object{sources=[]dto.SourceResponse,query=string}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /sources/article/{articleId}/search [get]
func SearchSimilarSources(c *fiber.Ctx) error {
	svc := getService()
	articleIDStr := c.Params("articleId")
	articleID, err := uuid.Parse(articleIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid article ID"))
	}

	query := c.Query("q", "")
	if query == "" {
		return response.Error(c, core.InvalidInputError("Query parameter 'q' is required"))
	}

	limit := c.QueryInt("limit", 5)
	if limit > 20 {
		limit = 20
	}

	sources, err := svc.SearchSimilar(c.Context(), articleID, query, limit)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"sources": sources,
		"query":   query,
	})
}

File: backend/src/api/source/handlers.rs
use axum::{
    Json,
    extract::{Path, Query, State, rejection::JsonRejection},
    http::StatusCode,
};
use uuid::Uuid;

use crate::api::{auth::AuthenticatedAccount, response::SuccessResponse};

use super::{
    dto::{
        CreateSourceRequest, ScrapeSourceRequest, SearchSourcesResponse, SourceListQuery,
        SourceListResponse, SourceResponse, SourceSearchQuery, SourcesResponse, SuccessFlag,
        UpdateSourceRequest,
    },
    error::SourceApiError,
    state::SourceState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, SourceApiError>;

#[utoipa::path(
    get,
    path = "/dashboard/sources",
    params(SourceListQuery),
    responses(
        (status = 200, body = SuccessResponse<SourceListResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "listAllSources"
)]
pub async fn list_all_sources(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    Query(query): Query<SourceListQuery>,
) -> ApiResult<SourceListResponse> {
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .list(query.page.unwrap_or(1), query.limit.unwrap_or(20))
            .await?
            .into(),
    )))
}

#[utoipa::path(
    post,
    path = "/sources",
    request_body = CreateSourceRequest,
    responses(
        (status = 201, body = SuccessResponse<SourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "createSource"
)]
pub async fn create_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    body: Result<Json<CreateSourceRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<SourceResponse>>), SourceApiError> {
    let Json(request) = body.map_err(invalid_request_body)?;
    let value = state
        .service()
        .create(request.validate().map_err(SourceApiError::validation)?)
        .await?;
    Ok((
        StatusCode::CREATED,
        Json(SuccessResponse::new(value.into())),
    ))
}

#[utoipa::path(
    post,
    path = "/sources/scrape",
    request_body = ScrapeSourceRequest,
    responses(
        (status = 201, body = SuccessResponse<SourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "scrapeAndCreateSource"
)]
pub async fn scrape_and_create_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    body: Result<Json<ScrapeSourceRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<SourceResponse>>), SourceApiError> {
    let Json(request) = body.map_err(invalid_request_body)?;
    request.validate().map_err(SourceApiError::validation)?;
    let value = state
        .service()
        .scrape_and_create(request.article_id, &request.url)
        .await?;
    Ok((
        StatusCode::CREATED,
        Json(SuccessResponse::new(value.into())),
    ))
}

#[utoipa::path(
    get,
    path = "/sources/article/{articleId}",
    params(("articleId" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SourcesResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "getArticleSources"
)]
pub async fn get_article_sources(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    Path(article_id): Path<String>,
) -> ApiResult<SourcesResponse> {
    let article_id = parse_article_id(&article_id)?;
    Ok(Json(SuccessResponse::new(SourcesResponse {
        sources: state
            .service()
            .get_by_article_id(article_id)
            .await?
            .into_iter()
            .map(Into::into)
            .collect(),
    })))
}

#[utoipa::path(
    get,
    path = "/sources/article/{articleId}/search",
    params(("articleId" = Uuid, Path), SourceSearchQuery),
    responses(
        (status = 200, body = SuccessResponse<SearchSourcesResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "searchSimilarSources"
)]
pub async fn search_similar_sources(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    Path(article_id): Path<String>,
    Query(query): Query<SourceSearchQuery>,
) -> ApiResult<SearchSourcesResponse> {
    let article_id = parse_article_id(&article_id)?;
    if query.q.is_empty() {
        return Err(crate::error::AppError::InvalidInput(
            "Query parameter 'q' is required".to_owned(),
        )
        .into());
    }
    let limit = query.limit.unwrap_or(5).min(20);
    let sources = state
        .service()
        .search_similar(article_id, &query.q, limit)
        .await?
        .into_iter()
        .map(Into::into)
        .collect();
    Ok(Json(SuccessResponse::new(SearchSourcesResponse {
        sources,
        query: query.q,
    })))
}

#[utoipa::path(
    get,
    path = "/sources/{sourceId}",
    params(("sourceId" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "getSource"
)]
pub async fn get_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    Path(source_id): Path<String>,
) -> ApiResult<SourceResponse> {
    let source_id = parse_source_id(&source_id)?;
    Ok(Json(SuccessResponse::new(
        state.service().get_by_id(source_id).await?.into(),
    )))
}

#[utoipa::path(
    put,
    path = "/sources/{sourceId}",
    params(("sourceId" = Uuid, Path)),
    request_body = UpdateSourceRequest,
    responses(
        (status = 200, body = SuccessResponse<SourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "updateSource"
)]
pub async fn update_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    Path(source_id): Path<String>,
    body: Result<Json<UpdateSourceRequest>, JsonRejection>,
) -> ApiResult<SourceResponse> {
    let source_id = parse_source_id(&source_id)?;
    let Json(request) = body.map_err(invalid_request_body)?;
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .update(source_id, request.into())
            .await?
            .into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/sources/{sourceId}",
    params(("sourceId" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "sources",
    operation_id = "deleteSource"
)]
pub async fn delete_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<SourceState>,
    Path(source_id): Path<String>,
) -> ApiResult<SuccessFlag> {
    let source_id = parse_source_id(&source_id)?;
    state.service().delete(source_id).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

fn parse_article_id(value: &str) -> Result<Uuid, SourceApiError> {
    Uuid::parse_str(value)
        .map_err(|_| crate::error::AppError::InvalidInput("Invalid article ID".to_owned()).into())
}

fn parse_source_id(value: &str) -> Result<Uuid, SourceApiError> {
    Uuid::parse_str(value)
        .map_err(|_| crate::error::AppError::InvalidInput("Invalid source ID".to_owned()).into())
}

fn invalid_request_body(_: JsonRejection) -> SourceApiError {
    crate::error::AppError::InvalidInput("Invalid request body".to_owned()).into()
}
